src/App/Options/Cmd/SqsToKafka.hs

```diff
@@ -64,34 +65,56 @@ instance RunApplication CmdSqsToKafka where
 
     runConduit $
       receiveMessageC sqsUrl
-      .| effectC (handleMessage sr kafkaTopic producer)
-      .| ackMessageC sqsUrl
+      .| batchByOrFlush (BatchSize 10)
+      .| effectC (handleMessages sr kafkaTopic producer)
+      .| ackMessages sqsUrl
       .| sinkNull
     return ()
 
-receiveMessageC :: MonadAWS m => String -> Source m Message
+receiveMessageC :: MonadAWS m => String -> Source m (Maybe Message)
 receiveMessageC sqsUrl = do
   let rm = receiveMessage (T.pack sqsUrl)
   rmr <- send (rm & (rmMaxNumberOfMessages .~ Just 10))
-  forM_ (rmr ^.. rmrsMessages . each) Conduit.yield
+  case rmr ^.. rmrsMessages . each of
+    []   -> yield Nothing
+    msgs -> forM_ msgs $ yield . Just
   receiveMessageC sqsUrl
 
-handleMessage :: (MonadIO m, MonadLogger m, MonadError AppError m) => SchemaRegistry -> TopicName -> KafkaProducer -> Message -> m ()
-handleMessage sr t@(TopicName topic) producer msg = do
-  fcm <- decodeSqsMessage msg & note (AppErr "Unable to decode SQS message") & eitherToError
-  payload <- encodeValue sr (Subject (T.pack topic)) fcm <&> left EncodeErr >>= eitherToError
-  let p = ProducerRecord t UnassignedPartition Nothing (Just (LBS.toStrict payload))
-  void $ produceMessage producer p
-  logInfo $ "Produced record: " <> show p
-
-ackMessageC :: (Monad m, MonadAWS m) => String -> Conduit Message m ()
-ackMessageC sqsUrl =
-  mapMC $ \msg -> do
-    let receipts = msg ^. mReceiptHandle
-
-    forM_ receipts $ \receipt -> do
-      let dm = deleteMessage (T.pack sqsUrl) receipt
-      void $ send dm
+handleMessages :: (MonadIO m, MonadLogger m, MonadError AppError m) => SchemaRegistry -> TopicName -> KafkaProducer -> [Message] -> m ()
+handleMessages sr t@(TopicName topic) producer msgs =
+  forM_ msgs $ \msg -> do
+    let sqsMessage = decodeSqsNotification msg
+    logInfo $ "sqs: " <> (show msg)
+
+    case sqsMessage of
+      Just (SqsMessageOfFileChangeMessage fcm) -> do
+        payload <- encodeValue sr (Subject (T.pack topic)) fcm <&> left EncodeErr >>= eitherToError
+        let p = ProducerRecord t UnassignedPartition Nothing (Just (LBS.toStrict payload))
+        void $ produceMessage producer p
+        logInfo $ "Produced record: " <> show p
+
+      Just SqsMessageOfS3TestEvent -> do
+        logInfo "s3:TestEvent found"
+        return ()
+
+      _ -> throwError (AppErr "Unable to decode SQS message")
+
+
+ackMessages :: (MonadAWS m, MonadError AppError m) => String -> Conduit [Message] m ()
+ackMessages sqsUrl =
+  mapMC $ \msgs -> do
+    let receipts = DM.catMaybes $ msgs ^.. each . mReceiptHandle
+    -- each dmbr needs an ID. just use the list index.
+    let dmbres = uncurry (\r i -> deleteMessageBatchRequestEntry (T.pack (show i)) r) <$> zip receipts ([0..] :: [Integer])
```

Comment:
Possible to replace

```
uncurry (\r i -> deleteMessageBatchRequestEntry (T.pack (show i)) r)
```

with

```
(\(r, i) -> deleteMessageBatchRequestEntry (T.pack (show i)) r)
```

?

Reply:
Arbitrary precision `Integer` is actually fairly slow.  Prefer `Int`.

Reply:
Is there something nicer than uncurry or tuple matching?

Reply:
Tuple matching is nice sometimes.